cclib/method/ddec.py

```diff
@@ -271,27 +272,94 @@ def calculate(self, indices=None, fupdate=0.05):
             self.radial_grid_r.append(r)
 
         # Carry out step 3 of DDEC6 algorithm [Determine conditioned charge density and tau]
-        self.logger.info("Conditioning charge densities.")
+        self.logger.info("Conditioning charge densities. (Step 3/7)")
         self.condition_densities()
 
+        # STEP 4-7
+        # Optimize weights on each grid point
+        self._kappa = [0.0 for x in self.data.atomnos]
+        update_kappa = False
+        self.N_A = []
+        self.u_A = []
+        steps = 4
+        cycle = 0
+        while steps < 8:
+            self.logger.info("Optimizing grid weights. (Step {}/7)".format(steps))
+            self.N_A.append(self._calculate_w_and_u())
+
+            # Update kappa as described in Figure S4.2 of doi: 10.1039/c6ra04656h
+            if numpy.any([x if x < -1e-5 else 0 for x in self.N_A[-1]]) and steps != 4:
+                update_kappa = True
+            if (
+                steps > 5
+                and update_kappa
+                and numpy.diff(self.N_A)[-1] < 1e-5
```

Comment:
So the definition of `numpy.diff` tells me this will actually be `self.N_A[0] - self.N_A[-1]` - is that what you wanted? Perhaps it's more clear to just write that out if the elements are scalars.

Reply:
Wouldn't this be `self.N_A[-1] - self.N_A[-2]`? I'll add a comment what this is. Thanks!